File: app/src/main/java/com/gmail/wondergab12/restapi/dependencies/ApplicationGraph.java
package com.gmail.wondergab12.restapi.dependencies;

import android.app.Application;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Singleton
@Component(modules = ApplicationModule.class)
public interface ApplicationGraph {

    @Component.Builder
    interface Builder {

        @BindsInstance
        Builder withApplication(Application application);

        ApplicationGraph build();
    }

    MainViewModelSubGraph.Builder mainViewModelSubGraphBuilder();

    MainActivitySubGraph.Builder mainActivitySubGraphBuilder();
}

File: app/src/main/java/com/gmail/wondergab12/restapi/dependencies/ApplicationGraph.kt
package com.gmail.wondergab12.restapi.dependencies

import android.app.Application
import dagger.BindsInstance
import dagger.Component

@ApplicationScope
@Component(modules = [ApplicationModule::class])
interface ApplicationGraph {

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun withApplication(application: Application): Builder

        fun build(): ApplicationGraph
    }

    fun mainViewModelSubGraphBuilder(): MainViewModelSubGraph.Builder

    fun mainActivitySubGraphBuilder(): MainActivitySubGraph.Builder
}
